CorsixTH/Lua/rooms/ward.lua

```diff
@@ -122,11 +122,29 @@ function WardRoom:doStaffUseCycle(humanoid)
   humanoid:queueAction(MeanderAction():setLoopCallback(meanders_loop))
 end
 
+-- This function is called once per tick per ward, unless there's no patient in the ward
+function WardRoom:getHealingAmount()
+  self.patient_idx = 0 -- Will be reset to nil after handling the last patient each tick
+  local nurse_force = 0 -- The total amount of work the nurses do
+  local best_staff = 0 -- Used for diagnosis quality
+  for humanoid in pairs(self.humanoids) do
+    if not humanoid:isLeaving() then
+      if class.is(humanoid, Patient) then
+        self.patient_idx = self.patient_idx + 1 -- Count the number of patients
+      elseif humanoid.humanoid_class == "Nurse" then
+        local nurse_quality = humanoid:getServiceQuality()
+        nurse_force = nurse_force + 1.0 + nurse_quality
+        if nurse_quality > best_staff then
+          best_staff = nurse_quality
+          self.staff_member = humanoid -- Used in \patient.lua\completeDiagnosticStep()
```

Comment:
So now the `completeDiagnosticStep` and `getHealingAmount` are intertwined in some complicated way. Please keep separate things separate. Let each function deal with its own computations.
Code is complicated enough already.

Reply:
It's finding the best nurse at the time, different to patient.lua which finds the one (and could find the best) staff member at the time of a diagnostic step. If one nurse leaves and a worse one comes in it should change the speed of progress? More relevant if you want progress to continue when there's no nurse in the room.

Reply:
My point is that you're counting patients, compute `nurse_force`, and compute `self.staff_member` all here now in one loop. This makes that all these things are now connected to each other, yet each value is independent. The number of patients doesn't change on `nurse_force`, both nurse values are not related to each other either.

By breaking this single computation in 3 separate computations (one for each result value), you're making the code more modular, and also show these values are really independent values.

The primary aim of code is to make it understandable to fellow human coders.

In the above loop anyone but you has to reverse-engineer how these variables are related currently (you're computing them together, the general assumption is that doing that has a reason.)
By splitting into several loops, you send the message "these things are independent". No need to decypher that information from the code, less chance of drawing wrong conclusions.